File: pkg/protocol/job.go
// Package protocol contains the transport types shared by SavewebHQ servers
// and Go clients. It deliberately contains no storage or networking code.
package protocol

import (
	"crypto/sha256"
	"encoding/hex"
)

const (
	JobTypeSeed  = "seed"
	JobTypeAsset = "asset"
)

// JobSpecV1 is the stable queue identity and work description exchanged by
// source tools, HQ, and workers.
type JobSpecV1 struct {
	ID    string         `json:"id,omitempty"`
	Value string         `json:"value"`
	Type  string         `json:"type,omitempty"`
	Via   *string        `json:"via,omitempty"`
	Hops  int            `json:"hops,omitempty"`
	Attrs map[string]any `json:"attr,omitempty"`
}

// DefaultJobID returns the v1 content-derived job ID. No value normalization is
// performed. An empty job type is treated as "seed".
func DefaultJobID(jobType, value string) string {
	if jobType == "" {
		jobType = JobTypeSeed
	}
	h := sha256.New()
	_, _ = h.Write([]byte(jobType))
	_, _ = h.Write([]byte{0})
	_, _ = h.Write([]byte(value))
	return "j1_" + hex.EncodeToString(h.Sum(nil))
}

File: sdk/python/src/saveweb_hq/job.py
"""Job identity helpers shared with the Go SDK."""

from hashlib import sha256


def default_job_id(job_type: str, value: str) -> str:
    """Return the v1 external job ID without normalizing *value*.

    An empty job type has the protocol default value ``seed``.
    """

    effective_type = job_type or "seed"
    digest = sha256(effective_type.encode() + b"\0" + value.encode()).hexdigest()
    return f"j1_{digest}"
